textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/common/RegexPredicate.java

```diff
@@ -1,88 +1,152 @@
-import java.util.regex.Matcher;
-import java.util.regex.Pattern;
+import java.util.stream.Collectors;
 
+import org.apache.lucene.analysis.Analyzer;
+import org.apache.lucene.queryparser.classic.MultiFieldQueryParser;
+import org.apache.lucene.queryparser.classic.ParseException;
+import org.apache.lucene.queryparser.classic.QueryParser;
+import org.apache.lucene.search.Query;
+
+import edu.uci.ics.textdb.api.common.Attribute;
+import edu.uci.ics.textdb.api.common.FieldType;
 import edu.uci.ics.textdb.api.common.IField;
 import edu.uci.ics.textdb.api.common.IPredicate;
 import edu.uci.ics.textdb.api.common.ITuple;
+import edu.uci.ics.textdb.api.dataflow.ISourceOperator;
+import edu.uci.ics.textdb.api.storage.IDataStore;
+import edu.uci.ics.textdb.common.constants.DataConstants;
+import edu.uci.ics.textdb.common.exception.DataFlowException;
 import edu.uci.ics.textdb.common.field.Span;
-import edu.uci.ics.textdb.common.field.StringField;
-import edu.uci.ics.textdb.common.field.TextField;
+import edu.uci.ics.textdb.dataflow.regexmatch.RegexToGramQueryTranslator;
+import edu.uci.ics.textdb.dataflow.source.IndexBasedSourceOperator;
+import edu.uci.ics.textdb.storage.DataReaderPredicate;
 
 /**
+ * This class is responsible for calculating regex match. <br>
+ * It uses RegexToGramQueryTranslator to translate a regex to a query in lucene. <br>
+ * 
+ * @author Zuozhi Wang
+ * @author Shuying Lai
  * @author sandeepreddy602
  *
  */
-public class RegexPredicate implements IPredicate{
+public class RegexPredicate implements IPredicate {
+
+	private String regex;
+	private Analyzer analyzer;
+	private List<String> fields;
+	private Query luceneQuery;
+	private ISourceOperator sourceOperator;
+	
+	private enum RegexEngine {
+		JavaRegex,
+		RE2J
+	}
+	private RegexEngine regexEngine;
+	
+
+	public RegexPredicate(String regex, List<Attribute> attributeList, Analyzer analyzer, IDataStore dataStore)
+			throws DataFlowException{
+		try {
+			this.regex = regex;
+			this.analyzer = analyzer;		
+			this.fields = attributeList.stream()
+					.filter(attr -> (attr.getFieldType() == FieldType.TEXT || attr.getFieldType() == FieldType.STRING))
+					.map(attr -> attr.getFieldName())
+					.collect(Collectors.toList());
+			
+			// try to use RE2J
+			try {
+				com.google.re2j.Pattern.compile(regex);
+				regexEngine = RegexEngine.RE2J;
+				this.luceneQuery = generateQuery(this.regex, this.fields, 
+						RegexToGramQueryTranslator.translate(this.regex).getLuceneQueryString());
+			// if RE2J failes, try to use Java Regex
+			} catch (com.google.re2j.PatternSyntaxException re2jException) {
+				java.util.regex.Pattern.compile(regex);
+				regexEngine = RegexEngine.JavaRegex;
+				this.luceneQuery = generateQuery(this.regex, this.fields, DataConstants.SCAN_QUERY);
+			}
+			
+			this.sourceOperator = new IndexBasedSourceOperator(new DataReaderPredicate(dataStore, luceneQuery));
+		} catch (ParseException | java.util.regex.PatternSyntaxException e) {

```

Comment:
Just curious: in Java can we import two packages with the same exception names?  I thought we can, since they belong to different packages, and we specify the full path to make them different?